### DISCUSSION TITLE
Calling a .Net Function from Javascript

### DISCUSSION THREAD

**Discussion by rachelYedatop** [cefsharp/CefSharp#4126] [General]
Hi @amaitland,
Thanks for all the help.
I tried the C # side to connect the Angular js (Angular1) side.
But it does not work for me. Maybe because of Js side? (to function: `closeApp()`)
C#:
 ```
browser.JavascriptObjectRepository.Register("jsObj", new ScriptManager(this), isAsync: false, options: BindingOptions.DefaultBinder);
 this.Controls.Add(browser);
 browser.LoadUrl("https://.........");
```
ScriptManager:
```
 private Form1 mForm;
  public ScriptManager(Form1 form)
  {
      mForm = form;
  }
```
closeApp  - the specific function:
```
public void closeApp()
{
    mForm.closeApp();
}
```
Js side:
`window.external.closeApp();` or `window.closeApp();`

**Comment by amaitland** [comment]
Firstly if you just want to close the window then calling https://developer.mozilla.org/en-US/docs/Web/API/Window/close should work by default in `WinForms`.

**Reply by rachelYedatop** [reply]
Ok, you're right,
I have a many .NET side functions that worked well with winform, 
but now with cefsharp I'm getting a little tangled...

**Reply by amaitland** [reply]
Your example only shows closing the form, so I always try to provide the simplest option. The code you've provided requires a number of changes.

- Use Async binding where possible
- The name you register the object with will be the object name in JavaScript.
- Legacy javascript binding automatically creates the object in each V8Context

Read https://github.com/cefsharp/CefSharp/wiki/General-Usage#3-how-do-you-expose-a-net-class-to-javascript for further information.

```c#
browser.JavascriptObjectRepository.Settings.LegacyBindingEnabled = true;
browser.JavascriptObjectRepository.Register("jsObj", new ScriptManager(this), isAsync: true, options: BindingOptions.DefaultBinder);
```

```jscript
window.jsObj.closeApp();
```

```c#
public void closeApp()
{
  mForm.BeginInvoke((Action)(() =>
  {
      mForm.closeApp();
  }));
}
```

**Reply by amaitland** [reply]
If you require assistance with porting from another browser then I'm happy to provide paid consulting services via [GitHub Sponsors](https://github.com/sponsors/amaitland)

**Reply by rachelYedatop** [reply]
Thanks @amaitland,
You really are a champion it works good.
